components/inputs/input-date-range.js

```diff
@@ -129,16 +149,62 @@ class InputDateRange extends RtlMixin(LocalizeCoreElement(LitElement)) {
 		if (input) input.focus();
 	}
 
+	async validate(showErrors) {
+		const errors = await super.validate(showErrors);
+		if (errors.length !== 0) {
+			return errors;
+		}
+		return Promise.all([
+			await this.shadowRoot.querySelector('.d2l-input-date-range-start').validate(showErrors),
+			await this.shadowRoot.querySelector('.d2l-input-date-range-end').validate(showErrors)]
+		).then((res) => {
+			return res.reduce((acc, errors) => [...acc, ...errors], []);
+		});
+	}
+
+	get validationMessageBadInput() {
+		return this.localize('components.input-date-range.errorBadInput', { startLabel: this._computedStartLabel, endLabel: this._computedEndLabel });
+	}
+
+	get _computedEndLabel() {
+		return this.endLabel ? this.endLabel : this.localize('components.input-date-range.endDate');
+	}
+
+	get _computedStartLabel() {
+		return this.startLabel ? this.startLabel : this.localize('components.input-date-range.startDate');
+	}
+
 	async _handleChange(e) {
-		// TODO: validation that start is before end
 		const elem = e.target;
-		if (elem.classList.contains('d2l-input-date-range-start')) this.startValue = elem.value;
-		else this.endValue = elem.value;
+		let startDate, endDate;
+		if (elem.classList.contains('d2l-input-date-range-start')) {
+			this.startValue = elem.value;
+			startDate = elem.value;
+			endDate = this.shadowRoot.querySelector('d2l-input-date.d2l-input-date-range-end').value;
+		} else {
+			this.endValue = elem.value;
+			endDate = elem.value;
+			startDate = this.shadowRoot.querySelector('d2l-input-date.d2l-input-date-range-start').value;
```

Comment:
Is there a reason we need to get this from the input rather than `this.startValue`?

Reply:
I think the way I'd originally implemented it needed it but no it doesn't need to be like this anymore - will fix.